File: data_structure/queue/src/QueueDemo.rb
# Ruby
# データ構造: キュー (Queue)

class QueueData
  def initialize
    @data = []
  end
  
  def get
    @data
  end

  def get_index(item)
    begin
      index = @data.index(item)
      return index
    rescue IndexError
      puts "ERROR: #{item} は範囲外です"
      -1
    end
  end

  def get_value(index)
    if index >= 0 && index < @data.length
      @data[index]
    else
      puts "Error: インデックス #{index} は範囲外です"
      nil
    end
  end

  def enqueue(item)
    @data.push(item)
    true
  end
  
  def dequeue
    unless is_empty?
      @data.shift
      true
    else
      puts "ERROR: キューが空です"
      false
    end
  end
  
  def peek
    unless is_empty?
      @data.first
    else
      puts "ERROR: キューが空です"
      nil
    end
  end
  
  def is_empty?
    @data.empty?
  end
  
  def size
    @data.length
  end
  
  def clear
    @data.clear
    true
  end
end

def main
  puts "Queue TEST -----> start"

  puts "\nnew"
  queue_data = QueueData.new
  puts "  現在のデータ: #{queue_data.get}"

  puts "\nis_empty"
  output = queue_data.is_empty?
  puts "  出力値: #{output}"

  puts "\nenqueue"
  input = [10, 20, 30]
  input.each do |item|
    puts "  入力値: #{item}"
    output = queue_data.enqueue(item)
    puts "  出力値: #{output}"
    puts "  現在のデータ: #{queue_data.get}"
  end

  puts "\nsize"
  output = queue_data.size
  puts "  出力値: #{output}"

  puts "\npeek"
  output = queue_data.peek
  puts "  出力値: #{output}"

  puts "\nget_index"
  input = 20
  puts "  入力値: #{input}"
  output = queue_data.get_index(input)
  puts "  出力値: #{output}"

  puts "\nget_index"
  input = 50
  puts "  入力値: #{input}"
  output = queue_data.get_index(input)
  puts "  出力値: #{output}"

  puts "\ndequeue"
  output = queue_data.dequeue
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{queue_data.get}"

  puts "\ndequeue"
  output = queue_data.dequeue
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{queue_data.get}"

  puts "\nsize"
  output = queue_data.size
  puts "  出力値: #{output}"

  puts "\ndequeue"
  output = queue_data.dequeue
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{queue_data.get}"

  puts "\ndequeue"
  output = queue_data.dequeue
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{queue_data.get}"

  puts "\nis_empty"
  output = queue_data.is_empty?
  puts "  出力値: #{output}"

  puts "\nclear"
  output = queue_data.clear
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{queue_data.get}"

  puts "\nsize"
  output = queue_data.size
  puts "  出力値: #{output}"

  puts "\nQueue TEST <----- end"
end

main if __FILE__ == $0

File: data_structure/queue/src/QueueDemo.java
// Java
// データ構造: キュー (Queue)

import java.util.LinkedList;
import java.util.Queue;

class QueueData {
    private Queue<Integer> data;

    public QueueData() {
        data = new LinkedList<>();
    }

    public Queue<Integer> get() {
        return data;
    }

    public int getIndex(Integer item) {
        Integer[] array = data.toArray(new Integer[0]);
        for (int i = 0; i < array.length; i++) {
            if (array[i].equals(item)) {
                return i;
            }
        }
        System.out.println("ERROR: " + item + " は範囲外です");
        return -1;
    }

    public Integer getValue(int index) {
        Integer[] array = data.toArray(new Integer[0]);
        if (0 <= index && index < array.length) {
            return array[index];
        } else {
            System.out.println("Error: インデックス " + index + " は範囲外です");
            return null;
        }
    }

    public boolean enqueue(Integer item) {
        return data.offer(item);
    }

    public boolean dequeue() {
        if (!isEmpty()) {
            data.poll();
            return true;
        } else {
            System.out.println("ERROR: キューが空です");
            return false;
        }
    }

    public Integer peek() {
        if (!isEmpty()) {
            return data.peek();
        } else {
            System.out.println("ERROR: キューが空です");
            return null;
        }
    }

    public boolean isEmpty() {
        return data.isEmpty();
    }

    public int size() {
        return data.size();
    }

    public boolean clear() {
        data.clear();
        return true;
    }
}

public class QueueDemo {
    public static void main(String[] args) {
        System.out.println("Queue TEST -----> start");

        System.out.println("\nnew");
        QueueData queueData = new QueueData();
        System.out.println("  現在のデータ: " + queueData.get());

        System.out.println("\nis_empty");
        boolean output = queueData.isEmpty();
        System.out.println("  出力値: " + output);

        System.out.println("\nenqueue");
        Integer[] input = {10, 20, 30};
        for (Integer item : input) {
            System.out.println("  入力値: " + item);
            output = queueData.enqueue(item);
            System.out.println("  出力値: " + output);
            System.out.println("  現在のデータ: " + queueData.get());
        }

        System.out.println("\nsize");
        int outputSize = queueData.size();
        System.out.println("  出力値: " + outputSize);

        System.out.println("\npeek");
        Integer outputPeek = queueData.peek();
        System.out.println("  出力値: " + outputPeek);

        System.out.println("\nget_index");
        int inputIndex = 20;
        System.out.println("  入力値: " + inputIndex);
        int outputIndex = queueData.getIndex(inputIndex);
        System.out.println("  出力値: " + outputIndex);

        System.out.println("\nget_index");
        inputIndex = 50;
        System.out.println("  入力値: " + inputIndex);
        outputIndex = queueData.getIndex(inputIndex);
        System.out.println("  出力値: " + outputIndex);

        System.out.println("\ndequeue");
        output = queueData.dequeue();
        System.out.println("  出力値: " + output);
        System.out.println("  現在のデータ: " + queueData.get());

        System.out.println("\ndequeue");
        output = queueData.dequeue();
        System.out.println("  出力値: " + output);
        System.out.println("  現在のデータ: " + queueData.get());

        System.out.println("\nsize");
        outputSize = queueData.size();
        System.out.println("  出力値: " + outputSize);

        System.out.println("\ndequeue");
        output = queueData.dequeue();
        System.out.println("  出力値: " + output);
        System.out.println("  現在のデータ: " + queueData.get());

        System.out.println("\ndequeue");
        output = queueData.dequeue();
        System.out.println("  出力値: " + output);
        System.out.println("  現在のデータ: " + queueData.get());

        System.out.println("\nis_empty");
        output = queueData.isEmpty();
        System.out.println("  出力値: " + output);

        System.out.println("\nclear");
        output = queueData.clear();
        System.out.println("  出力値: " + output);
        System.out.println("  現在のデータ: " + queueData.get());

        System.out.println("\nsize");
        outputSize = queueData.size();
        System.out.println("  出力値: " + outputSize);

        System.out.println("\nQueue TEST <----- end");
    }
}